clients/src/main/java/org/apache/kafka/common/security/auth/DefaultKafkaPrincipalBuilder.java

```
@@ -0,0 +1,107 @@
+    public KafkaPrincipal build(AuthenticationContext context) {
+        if (context instanceof PlaintextAuthenticationContext) {
+            if (oldPrincipalBuilder != null)
+                return convertToKafkaPrincipal(oldPrincipalBuilder.buildPrincipal(transportLayer, authenticator));
+
+            return KafkaPrincipal.ANONYMOUS;
+        } else if (context instanceof SslAuthenticationContext) {
+            SSLSession sslSession = ((SslAuthenticationContext) context).session();
+
+            if (oldPrincipalBuilder != null)
+                return convertToKafkaPrincipal(oldPrincipalBuilder.buildPrincipal(transportLayer, authenticator));
+
+            try {
+                return convertToKafkaPrincipal(sslSession.getPeerPrincipal());
+            } catch (SSLPeerUnverifiedException se) {
+                return KafkaPrincipal.ANONYMOUS;
+            }
+        } else if (context instanceof SaslAuthenticationContext) {
+            SaslServer saslServer = ((SaslAuthenticationContext) context).server();
+            if (SaslConfigs.GSSAPI_MECHANISM.equals(saslServer.getMechanismName()))
+                return applyKerberosShortNamer(saslServer.getAuthorizationID());
+            else
+                return new KafkaPrincipal(KafkaPrincipal.USER_TYPE, saslServer.getAuthorizationID());
+        } else {
+            throw new IllegalArgumentException("Unhandled authentication context type: " + context.getClass().getName());
+        }
+    }
+
+    private KafkaPrincipal applyKerberosShortNamer(String authorizationId) {
+        KerberosName kerberosName = KerberosName.parse(authorizationId);
+        try {
+            String shortName = kerberosShortNamer.shortName(kerberosName);
+            return new KafkaPrincipal(KafkaPrincipal.USER_TYPE, shortName);
+        } catch (IOException e) {
+            log.error("Failed to set name for '{}' based on Kerberos authentication rules.", kerberosName, e);
+            return KafkaPrincipal.ANONYMOUS;
```

Should we not propagate this exception?

---

Hmm.. that's fair. In the old implementation, we ignored the exception and used the raw authenticationId. I guess we could do the same here, but maybe it's better to propagate the exception?

---

Propagating seems better to me.